external/llvm-project/mlir/lib/Transforms/Utils/OutlinerUtils.cpp

```diff
@@ -173,337 +209,380 @@ static bool opsMatch(Operation *lhs, Operation *rhs, OutliningCandidate &one,
+    if (inOp == nullptr)
+      return false;
+    else if (inOp == _anchorOp)
+      return true;
+    else if (encountered.contains(inOp))
       return false;
+    encountered.insert(inOp);
+    for (auto opr : inOp->getOperands()) {
+      if (Operation *oprOp = opr.getDefiningOp()) {
+        debug(">>>", oprOp);
+        if (inChainToAnchor(oprOp, encountered))
+          return true;
+        debug("<<<", oprOp);
+      }
     }
+    return false;
   }
-  return true;
-}
 
-static OutliningCandidate *
-findOutliningCandidate(OutliningCandidate &newCandidate,
-                       std::vector<OutliningCandidate> &candidates) {
-  for (auto &candidate : candidates) {
-    if (outliningCandidatesEquivalent(candidate, newCandidate)) {
-      return &candidate;
+  // Make sure all operands are external or immediate in the chain from anchorOp
+  bool isImmediateOp(Operation *inOp) const {
+    for (auto opr : inOp->getOperands()) {
+      if (Operation *oprOp = opr.getDefiningOp()) {
+        debug("IN>", oprOp);
+        DenseSet<Operation *> encountered;
+        if (!contains(oprOp) && inChainToAnchor(oprOp, encountered))
+          return false;
+      }
     }
+    return true;
   }
-  return nullptr;
-}
 
-// Given an op and its fuse-able trailing (second) and leading
-// (front) ops, remove them into a separate function.
-static void outlineOps(Operation *anchorOp, ArrayRef<Operation *> trailingOps,
-                       ArrayRef<Operation *> leadingOps, ArrayRef<Value> params,
-                       ArrayRef<Value> returnVals, StringRef partFnName,
-                       StringRef attrName,
-                       std::vector<OutliningCandidate> &candidates) {
-  ValueRange values(params);
-  OpBuilder b(anchorOp);
-  Location loc = anchorOp->getLoc();
-  func::FuncOp outlinedFunc;
-  Location fusedLoc(loc);
-
-  // ------------------------------------------------------------
-  // Merging part.
-
-  OutliningCandidate newCandidate(anchorOp, trailingOps, leadingOps, params,
-                                  returnVals, partFnName);
-
-  if (OutliningCandidate *found =
-          findOutliningCandidate(newCandidate, candidates)) {
-    // Matches one we already have.
-    outlinedFunc = found->function;
-    fusedLoc = found->fusedLoc;
-  } else {
-    // ------------------------------------------------------------
-    // Construction part.
+  // Recurse all inputs to op and capture predecessor ops if qualified
+  void collectInputs(Operation *op, Operation *ignoreOp) {
+    for (const auto &operand : op->getOperands()) {
+      Operation *usedOp = operand.getDefiningOp();
+      if (usedOp) {
+        // Ignore input from anchor chain or already collected
+        if (usedOp != ignoreOp && !contains(usedOp)) {
+          if (_outliner.isLeadingOp(usedOp)) {
+            // Depth first collection of Leading ops
+            collectInputs(usedOp, op);
+            addPredecessor(usedOp);
```

Comment:
The reason for the little remove-and-insert dance that my version did is this situation:

A
|\
| - op2
|    |
 \   |
  - op1
     |
     |
   anchor

When collecting the inputs of op1, we see A first and add it to the leading ops.  Then we visit op1 and add it, and it also has A as an operand, but they're in the wrong order and a later stage will complain.  I had this happen but don't recall the example.  I can't convince myself that doing recur-then-add as you have here will avoid the problem.

Reply:
Yes, I made leadingOps add in order of encounter, so I reversed the order here of recursion to go recurse and add all inputs and ops before adding the current op.